modules/calib3d/src/calibration.cpp

```diff
@@ -59,460 +58,38 @@
 using namespace cv;
 
 // reimplementation of dAB.m
-CV_IMPL void cvCalcMatMulDeriv( const CvMat* A, const CvMat* B, CvMat* dABdA, CvMat* dABdB )
+CV_IMPL void cvCalcMatMulDeriv( const CvMat* _A, const CvMat* _B, CvMat* _dABdA, CvMat* _dABdB )
 {
-    int i, j, M, N, L;
-    int bstep;
-
-    CV_Assert( CV_IS_MAT(A) && CV_IS_MAT(B) );
-    CV_Assert( CV_ARE_TYPES_EQ(A, B) &&
-        (CV_MAT_TYPE(A->type) == CV_32F || CV_MAT_TYPE(A->type) == CV_64F) );
-    CV_Assert( A->cols == B->rows );
-
-    M = A->rows;
-    L = A->cols;
-    N = B->cols;
-    bstep = B->step/CV_ELEM_SIZE(B->type);
-
-    if( dABdA )
-    {
-        CV_Assert( CV_ARE_TYPES_EQ(A, dABdA) &&
-            dABdA->rows == A->rows*B->cols && dABdA->cols == A->rows*A->cols );
-    }
-
-    if( dABdB )
-    {
-        CV_Assert( CV_ARE_TYPES_EQ(A, dABdB) &&
-            dABdB->rows == A->rows*B->cols && dABdB->cols == B->rows*B->cols );
-    }
-
-    if( CV_MAT_TYPE(A->type) == CV_32F )
-    {
-        for( i = 0; i < M*N; i++ )
-        {
-            int i1 = i / N,  i2 = i % N;
-
-            if( dABdA )
-            {
-                float* dcda = (float*)(dABdA->data.ptr + dABdA->step*i);
-                const float* b = (const float*)B->data.ptr + i2;
-
-                for( j = 0; j < M*L; j++ )
-                    dcda[j] = 0;
-                for( j = 0; j < L; j++ )
-                    dcda[i1*L + j] = b[j*bstep];
-            }
-
-            if( dABdB )
-            {
-                float* dcdb = (float*)(dABdB->data.ptr + dABdB->step*i);
-                const float* a = (const float*)(A->data.ptr + A->step*i1);
-
-                for( j = 0; j < L*N; j++ )
-                    dcdb[j] = 0;
-                for( j = 0; j < L; j++ )
-                    dcdb[j*N + i2] = a[j];
-            }
-        }
-    }
-    else
-    {
-        for( i = 0; i < M*N; i++ )
-        {
-            int i1 = i / N,  i2 = i % N;
-
-            if( dABdA )
-            {
-                double* dcda = (double*)(dABdA->data.ptr + dABdA->step*i);
-                const double* b = (const double*)B->data.ptr + i2;
-
-                for( j = 0; j < M*L; j++ )
-                    dcda[j] = 0;
-                for( j = 0; j < L; j++ )
-                    dcda[i1*L + j] = b[j*bstep];
-            }
-
-            if( dABdB )
-            {
-                double* dcdb = (double*)(dABdB->data.ptr + dABdB->step*i);
-                const double* a = (const double*)(A->data.ptr + A->step*i1);
-
-                for( j = 0; j < L*N; j++ )
-                    dcdb[j] = 0;
-                for( j = 0; j < L; j++ )
-                    dcdb[j*N + i2] = a[j];
-            }
-        }
-    }
+    Mat A = cvarrToMat(_A), B = cvarrToMat(_B);
+    Mat dABdA = cvarrToMat(_dABdA), dABdB = cvarrToMat(_dABdB);
```

Comment:
`dABdA` and `dABdB` are optional. `cvarrToMat` returns empty Mat if the pointer is nullptr. Empty Mat is not equal to `cv::NoArray()`. You have to check pointers on wrapper side and put `cv::noArray` for optional outputs.